### PR TITLE
move2kube workflow using ssh keys instead of token

### PR DESCRIPTION
**Pull request by gabriel-farache**
Follow-up of #8 

This PR changes the way the workflow and the move2kube instance interact with the private git repo: instead of using a token, they are using ssh keys previously added to the account (ie: https://bitbucket.org/account/settings/ssh-keys/) 

By using the ssh keys, the move2kube instance can clone the repository when creating a plan. Thus, enabling the workflow to use the `remote-source` query param and removing the need for the `CreatePlan` Knative function.

Once https://github.com/konveyor/move2kube/pull/1126 is merged, we can revert back to using the `konveyor` images

Solves https://issues.redhat.com/browse/FLPATH-820

### CODE REVIEW

File: m2k/README.md [hunk 1]
```diff
@@ -169,16 +182,20 @@ Should output
 quay.io/orchestrator/serverless-workflow-m2k                                    <none>         cd2e0498ee70   4 minutes ago   487MB
 ```
 #### 4. M2K Knative functions and GC
-* [m2k-service.yaml](k8s/m2k-service.yaml) will deploy 2 kservices that will spin-up the functions when an event is received
+* [m2k-service.yaml](k8s/m2k-service.yaml) will deploy the Knative service that will spin-up the functions when an event is received
 * [m2k-trigger.yaml](k8s/m2k-trigger.yaml) will deploy the triggers related to the expected event and to which the kservice subscribes and rely on to get started
 * [knative-gc.yaml](k8s%2Fknative-gc.yaml) will setup the GC to keep only 3 revisions in the cluster
-From the root folder of the project, first create the Knative services:
+
+As we are using ssh keys to interact with the private git repo (ie: bitbucket), similarly to what we have done when deploying the `move2kube` instance, we need to create secrets in the `m2k` namespace containing the keys:
```
Comment (pkliczewski) [inline m2k/README.md:189]: we are using the keys to interact with any repo. please remove `private`

File: m2k/k8s/m2k-service.yaml [hunk 2]
```diff
@@ -1,42 +1,54 @@
...
+          volumeMounts:
+            - name: ssh-priv-key
+              mountPath: "/root/.ssh/id_rsa"
+              subPath: id_rsa
+              readOnly: true
+            - name: pre-install
+              mountPath: /etc/pre-install
       containers:
-        - image: quay.io/orchestrator/m2k-kfunc:latest
+        - image: quay.io/orchestrator/m2k-kfunc:2.0.0-SNAPSHOT
```
Comment (pkliczewski) [inline m2k/k8s/m2k-service.yaml:22]: we are not maintaining latest image anymore?

File: m2k/k8s/move2kube.yaml [hunk 3]
```diff
@@ -13,9 +13,33 @@ spec:
     spec:
       containers:
         - name: move2kube
-          image: quay.io/konveyor/move2kube-ui:latest
+          image: quay.io/orchestrator/move2kube-ui:latest
```
Comment (pkliczewski) [inline m2k/k8s/move2kube.yaml:16]: is there any issue with image maintained by konveyor team? if so it would be great to let the know about it

File: m2k/m2k-func/README.md [hunk 4]
```diff
@@ -1,18 +1,11 @@
...
-  * Then it will upload this archive to the requested Move2kube project under the provided workspace in order to create a planning
-  * Will send events:
-    * `plan_created_event_type` if success
-    * `error_event_type` if any error
 * SaveTransformationOutput:
   * Triggered by the event `save-transformation`
   * This function will first retrieve the transformation output archive from the Move2Kube project
   * Then it will create a new branch based on the provided input in the provided BitBucket repo
-  * Finally, it will un-archive the previously downloaded file, commit the change and push them to BitBucket using the provided token
+  * Finally, it will un-archive the previously downloaded file, commit the change and push them to BitBucket using the provided token if provided, otherwise the ssh keys will be used.
```
Comment (pkliczewski) [inline m2k/m2k-func/README.md:8]: I would rephrase. We use twice `provided` word. What about:

using the token if provided, otherwise the ssh keys will be used.

File: m2k/serverless-workflow-m2k/src/main/resources/application-knative.properties [hunk 5]
```diff
@@ -11,6 +11,7 @@ quarkus.container-image.build=true
 #quarkus.container-image.registry=dev.local
 quarkus.container-image.group=orchestrator
 quarkus.container-image.name=serverless-workflow-m2k
+quarkus.container-image.tag=2.0.0-SNAPSHOT
```
Comment (pkliczewski) [inline m2k/serverless-workflow-m2k/src/main/resources/application-knative.properties:14]: Why do we need to add it here? it this related to not using `latest` tag?

Comment (gabriel-farache) [reply m2k/serverless-workflow-m2k/src/main/resources/application-knative.properties:14]: To avoid overriding the one currently deployed and using token, which is `1.0.0-SNAPSHOT` currently, I'll create a `latest` that maps to the one on the `main` branch`

File: m2k/k8s/m2k-service.yaml (hunk 2 above)
Comment (gabriel-farache) [reply m2k/k8s/m2k-service.yaml:22]: To avoid overriding the one currently deployed and using token, I push the current work with a snapshot version, could be 1.0.0 but it was more in the idea of matching the same as for the serverless workflow image

File: m2k/k8s/move2kube.yaml (hunk 3 above)
Comment (gabriel-farache) [reply m2k/k8s/move2kube.yaml:16]: Yep, waiting on https://github.com/konveyor/move2kube/pull/1126 to be merged

File: m2k/README.md (hunk 1 above)
Comment (gabriel-farache) [reply m2k/README.md:189]: nope, we need to provide the private key to clone/push otherwise we get
```
2024-01-04 13:28:55,605 ERROR [dev.par.SaveTransformationFunction] (executor-thread-1) Cannot clone repo ssh://bitbucket.org/gfarache31/m2k-test: org.eclipse.jgit.api.errors.TransportException: ssh://git@bitbucket.org/gfarache31/m2k-test: invalid privatekey
        at org.eclipse.jgit.api.FetchCommand.call(FetchCommand.java:249)
        at org.eclipse.jgit.api.CloneCommand.fetch(CloneCommand.java:325)
        at org.eclipse.jgit.api.CloneCommand.call(CloneCommand.java:191)
        at dev.parodos.service.GitServiceImpl.cloneRepo(GitServiceImpl.java:51)
        at dev.parodos.service.GitServiceImpl_ClientProxy.cloneRepo(Unknown Source)
        at dev.parodos.SaveTransformationFunction.pushTransformationToGitRepo(SaveTransformationFunction.java:71)
        at dev.parodos.SaveTransformationFunction.saveTransformation(SaveTransformationFunction.java:63)
        at dev.parodos.SaveTransformationFunction_ClientProxy.saveTransformation(Unknown Source)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at io.quarkus.funqy.runtime.FunctionInvoker.invoke(FunctionInvoker.java:123)
        at io.quarkus.funqy.runtime.bindings.knative.events.VertxRequestHandler.dispatch(VertxRequestHandler.java:571)
        at io.quarkus.funqy.runtime.bindings.knative.events.VertxRequestHandler.lambda$processCloudEvent$4(VertxRequestHandler.java:404)
        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.eclipse.jgit.errors.TransportException: ssh://git@bitbucket.org/gfarache31/m2k-test: invalid privatekey
        at org.eclipse.jgit.transport.ssh.jsch.JschConfigSessionFactory.getSession(JschConfigSessionFactory.java:169)
        at org.eclipse.jgit.transport.SshTransport.getSession(SshTransport.java:107)
        at org.eclipse.jgit.transport.TransportGitSsh$SshFetchConnection.<init>(TransportGitSsh.java:279)
        at org.eclipse.jgit.transport.TransportGitSsh.openFetch(TransportGitSsh.java:152)
        at org.eclipse.jgit.transport.FetchProcess.executeImp(FetchProcess.java:153)
        at org.eclipse.jgit.transport.FetchProcess.execute(FetchProcess.java:105)
        at org.eclipse.jgit.transport.Transport.fetch(Transport.java:1465)
        at org.eclipse.jgit.api.FetchCommand.call(FetchCommand.java:238)
        ... 21 more
Caused by: com.jcraft.jsch.JSchException: invalid privatekey
        at com.jcraft.jsch.KeyPair.load(KeyPair.java:862)
        at com.jcraft.jsch.KeyPair.load(KeyPair.java:668)
```
It is what was done in parodos FYI

I'll check why we have to use the private key instead of the public key

File: m2k/m2k-func/README.md (hunk 4 above)
Comment (gabriel-farache) [reply m2k/m2k-func/README.md]: done

File: m2k/README.md (hunk 1 above)
Comment (pkliczewski) [reply m2k/README.md:189]: it is not what I meant. You wrote `the private git repo`. I think it will work for public repos as well hence `private` not needed

File: m2k/k8s/m2k-service.yaml (hunk 2 above)
Comment (pkliczewski) [reply m2k/k8s/m2k-service.yaml:22]: ok, I think we need to come up with some good practices how to handle function image versioning

File: m2k/README.md (hunk 1 above)
Comment (gabriel-farache) [reply m2k/README.md:189]: ah! yes ok

File: m2k/k8s/move2kube.yaml (hunk 3 above)
Comment (pkliczewski) [reply m2k/k8s/move2kube.yaml:16]: ok, thanks

File: m2k/serverless-workflow-m2k/src/main/resources/application-knative.properties (hunk 5 above)
Comment (pkliczewski) [reply m2k/serverless-workflow-m2k/src/main/resources/application-knative.properties:14]: makes sense. Do you think it would make sense to provide some tooling for workflow developers to update versions in so many places?

**Review by pkliczewski — COMMENTED** [review]
overall it looks good, please just fix this last comment

**Comment by pkliczewski** [comment]
@masayag please review and merge

**Review by masayag — APPROVED** [review]
Well-done!
